File: csharp/2015/day06a/Program.cs
using System;
using System.IO;
using System.Text.RegularExpressions;

using AOC_Utils.Geometry;

class Bounds
{
    public Position2D<int> UpperLeft;
    public Position2D<int> LowerRight;

    public Bounds(Position2D<int> upper, Position2D<int> lower)
    {
        UpperLeft = upper;
        LowerRight = lower;
    }
}

class Grid
{
    bool[,] grid;

    public const int X_MAX = 1000;
    public const int Y_MAX = 1000;

    public Grid()
    {
        grid = new bool[X_MAX, Y_MAX];
        for ( int j = 0; j < Y_MAX; j++ )
        {
            for ( int i = 0; i < X_MAX; i++ )
            {
                grid[j, i] = false;
            }
        }
    }

    public void Perform(string action, Bounds bounds)
    {
        for ( int j = bounds.UpperLeft.Y; j <= bounds.LowerRight.Y; j++ )
        {
            for ( int i = bounds.UpperLeft.X; i <= bounds.LowerRight.X; i++ )
            {
                switch ( action )
                {
                    case "turn on":
                        grid[j, i] = true;
                        break;
                    case "turn off":
                        grid[j, i] = false;
                        break;
                    case "toggle":
                        grid[j, i] = !grid[j, i];
                        break;
                }
            }
        }
    }

    public int Count()
    {
        int count = 0;
        for ( int j = 0; j < Y_MAX; j++ )
        {
            for ( int i = 0; i < X_MAX; i++ )
            {
                if ( grid[j, i] )
                {
                    count++;
                }
            }
        }
        return count;
    }
}

class Program
{
    static void Usage()
    {
        Console.Error.WriteLine("usage: dotnet run <input file>");
        Environment.Exit(1);
    }

    static int Process(string filename)
    {
        Grid grid = new Grid();
        Regex re = new Regex(@"(turn on|turn off|toggle) (\d+),(\d+) through (\d+),(\d+)");

        try
        {
            using ( StreamReader reader = new StreamReader(filename) )
            {
                string? line;
                while ( (line = reader.ReadLine()) != null )
                {
                    Match match = re.Match(line);
                    if ( match.Success )
                    {
                       string action = match.Groups[1].Value;
                       int.TryParse(match.Groups[2].Value, out int x1);
                       int.TryParse(match.Groups[3].Value, out int y1);
                       Position2D<int> upperLeft = new Position2D<int>(x1, y1);
                       int.TryParse(match.Groups[4].Value, out int x2);
                       int.TryParse(match.Groups[5].Value, out int y2);
                       Position2D<int> lowerRight = new Position2D<int>(x2, y2);
                       Bounds bounds = new Bounds(upperLeft, lowerRight);
                       grid.Perform(action, bounds);
                    }
                }
            }
        }
        catch ( IOException e )
        {
            Console.Error.WriteLine("Error reading file: " + e.Message);
            Environment.Exit(1);
        }

        return grid.Count();
    }

    static void Main(string[] args)
    {
        if ( args.Length < 1 )
        {
            Usage();
        }

        string filename = args[0];
        int result = Process(filename);
        Console.WriteLine("result = " + result);
    }
}

File: java/2015/day06a/Program.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.regex.*;

public class Program {
    static final int ROW_MAX = 1000;
    static final int COL_MAX = 1000;

    public static void usage() {
        System.err.println("usage: java -jar Program.jar <input file>");
        System.exit(1);
    }

    public static void perform(boolean[] grid, String action, int r1, int c1, int r2, int c2) {
        int row, col;

        for ( row = r1; row <= r2; row++ ) {
            for ( col = c1; col <= c2; col++ ) {
                switch ( action ) {
                case "turn on":
                    grid[row*COL_MAX + col] = true;
                    break;
                case "turn off":
                    grid[row*COL_MAX + col] = false;
                    break;
                case "toggle":
                    grid[row*COL_MAX + col] = !grid[row*COL_MAX + col];
                    break;
                default:
                    break;
                }
            }
        }
    }

    public static int count(boolean[] grid) {
        int row, col;
        int total = 0;

        for ( row = 0; row < ROW_MAX; row++ ) {
            for ( col = 0; col < COL_MAX; col++ ) {
                if ( grid[row*COL_MAX + col] ) {
                    total++;
                }
            }
        }
        return total;
    }

    public static int process(String filename) {
        Pattern p = Pattern.compile("(turn on|turn off|toggle) (\\d+),(\\d+) through (\\d+),(\\d+)");
        boolean[] grid = new boolean[ROW_MAX*COL_MAX];
        try ( BufferedReader reader = new BufferedReader(new FileReader(filename))) {
            String line;
            while ( (line = reader.readLine()) != null ) {
                Matcher m = p.matcher(line);
                if ( m.matches() ) {
                    String action = m.group(1);
                    int r1 = Integer.parseInt(m.group(2));
                    int c1 = Integer.parseInt(m.group(3));
                    int r2 = Integer.parseInt(m.group(4));
                    int c2 = Integer.parseInt(m.group(5));
                    perform(grid, action, r1, c1, r2, c2);
                }
            }
        } catch ( IOException e ) {
            System.err.println("Error reading file: " + e.getMessage());
            System.exit(1);
        }
        return count(grid);
    }

    public static final void main(String[] args) {
        if ( args.length < 1 ) {
            usage();
        }
        String filename = args[0];
        int result = process(filename);
        System.out.println("result = " + result);
    }
}
